CONFIGURATION-846 restore previous behavior allowing spring to inject multiple values

abollini opened the pull request:
This is a fix for [CONFIGURATION-846](https://issues.apache.org/jira/browse/CONFIGURATION-846)

The PR includes a specific test for the multiple values in spring, moreover I have verified that with this change the issue in our open source project where we noted the issue is solved

garydgregory commented on src/main/java/org/apache/commons/configuration2/spring/ConfigurationPropertySource.java:
```diff
@@ -39,7 +39,8 @@ public ConfigurationPropertySource(final String name, final Configuration source
     @Override
     public Object getProperty(final String name) {
-        return source.getString(name);
+        String[] propValue = source.getStringArray(name);
```
Use `final` where you can.

garydgregory reviewed (changes requested):
A minor comment but we need to make sure we are not going to ping-pong between behaviors with https://github.com/apache/commons-configuration/pull/309/files which has been merged already. CC @kbarlowgw

tdonohue commented:
@kbarlowgw and @garydgregory : Just for clarification, @abollini  is a collaborator of mine on the DSpace project where we discovered that [CONFIGURATION-846](https://issues.apache.org/jira/browse/CONFIGURATION-846) was caused by #309.  

I talked with @abollini today about this fix & he believes this refactor fixes [CONFIGURATION-846](https://issues.apache.org/jira/browse/CONFIGURATION-846) *without breaking* the fixes to [CONFIGURATION-834](https://issues.apache.org/jira/browse/CONFIGURATION-834) applied by #309.   

However, we'd very much appreciate it if @kbarlowgw can verify this PR doesn't break or reintroduce the issues found in CONFIGURATION-834.

abollini commented:
hi @garydgregory thanks for looking to that.
I have updated the PR to flag the variable as final as by your suggestion. I confirm that the expectation is to have preserved the goal of the fix for CONFIGURATION-834
Indeed, the test introduced by @kbarlowgw is still there
https://github.com/apache/commons-configuration/pull/425/files#diff-b3d5f2d39a771e588b33d0d9719c4edcbbdaf3059bca13b4f70d7e0226d74ee8R106

I have only updated the test value to make the test more reliable, moreover in the test that I have added for the multiple values the interpolation is also checked
https://github.com/apache/commons-configuration/pull/425/files#diff-b3d5f2d39a771e588b33d0d9719c4edcbbdaf3059bca13b4f70d7e0226d74ee8R115

garydgregory commented:
I would like to hear from @kbarlowgw and if this PR affects the use case from #309 negatively...

kbarlowgw commented:
Sorry for the delay... Unfortunately I did encounter an error (full error
attached):

    "loggerName":
"org.springframework.boot.diagnostics.LoggingFailureAnalysisReporter",
    "message": "Application failed to start due to an exception",
    "thrown":
    {
        "commonElementCount": 0,
        "localizedMessage": "Failed to bind properties under
'server.compression.enabled' to boolean",
        "message": "Failed to bind properties under
'server.compression.enabled' to boolean",
        "name":
"org.springframework.boot.context.properties.bind.BindException",
        "cause":
        {
            "commonElementCount": 94,
            "localizedMessage": "Failed to convert from type
[java.lang.String[]] to type [boolean] for value [null]",
            "message": "Failed to convert from type [java.lang.String[]] to
type [boolean] for value [null]",
            "name":
"org.springframework.core.convert.ConversionFailedException",
            "cause":
            {
                "commonElementCount": 94,
                "localizedMessage": "A null value cannot be assigned to a
primitive type",
                "message": "A null value cannot be assigned to a primitive
type",
                "name": "java.lang.IllegalArgumentException",
                "extendedStackTrace":
                [

I think what you want is:

@Override
public Object getProperty(final String name) {
  final String[] propValue = source.getStringArray(name);
  if (propValue == null || propValue.length == 0) {
    return null;
  } else if (propValue.length == 1) {
    return propValue[0];
  } else return propValue;
}


On Thu, May 23, 2024 at 1:26 PM Gary Gregory ***@***.***>
wrote:

> I would like to hear from @kbarlowgw
> <https://github.com/kbarlowgw>
> and if this PR affects the use case from #309
> <https://github.com/apache/commons-configuration/pull/309>
> negatively...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-configuration/pull/425#issuecomment-2127691962>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBG2EVFNN74NSRTDSHQMI3TZDYREVAVCNFSM6AAAAABID3UVXCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRXGY4TCOJWGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

abollini commented:
hi @kbarlowgw , 
thanks for your feedback. If I understood correctly I should have added a test related to the wrong behavior introduced dealing with unset property https://github.com/apache/commons-configuration/pull/425/commits/e0edce6f3676081d8efe3c2adf446025452bcf74

Indeed, running the test without further change I got the following exception similar to your
```
[ERROR] Errors: 
[ERROR]   TestConfigurationPropertySource.testListValueInjection » UnsatisfiedDependency Error creating bean with name 'org.apache.commons.configuration2.spring.TestConfigurationPropertySource': Unsatisfied dependency expressed through field 'booleanNullValue'; nested exception is org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'boolean'; nested exception is java.lang.IllegalArgumentException: Invalid boolean value []
[ERROR]   TestConfigurationPropertySource.testNullValueInjection » UnsatisfiedDependency Error creating bean with name 'org.apache.commons.configuration2.spring.TestConfigurationPropertySource': Unsatisfied dependency expressed through field 'booleanNullValue'; nested exception is org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'boolean'; nested exception is java.lang.IllegalArgumentException: Invalid boolean value []
[ERROR]   TestConfigurationPropertySource.testSystemPropertyValueInjection » UnsatisfiedDependency Error creating bean with name 'org.apache.commons.configuration2.spring.TestConfigurationPropertySource': Unsatisfied dependency expressed through field 'booleanNullValue'; nested exception is org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'boolean'; nested exception is java.lang.IllegalArgumentException: Invalid boolean value []
[ERROR]   TestConfigurationPropertySource.testValueInjection » UnsatisfiedDependency Error creating bean with name 'org.apache.commons.configuration2.spring.TestConfigurationPropertySource': Unsatisfied dependency expressed through field 'booleanNullValue'; nested exception is org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'boolean'; nested exception is java.lang.IllegalArgumentException: Invalid boolean value []
[INFO] 
[ERROR] Tests run: 2942, Failures: 0, Errors: 4, Skipped: 2
```

that are solved with the commit https://github.com/apache/commons-configuration/pull/425/commits/ab388c3a0c97cd249dfc2945b149da8b34e01da6

I would appreciate a new check on your side. Thanks!

kbarlowgw commented:
Yes, I think this works alright for me.

Thanks.

Keith

On Mon, May 27, 2024 at 5:25 AM Andrea Bollini ***@***.***>
wrote:

> hi @kbarlowgw
> <https://github.com/kbarlowgw>
> ,
> thanks for your feedback. If I understood correctly I should have added a
> test related to the wrong behavior introduced dealing with unset property
> e0edce6
> <https://github.com/apache/commons-configuration/commit/e0edce6f3676081d8efe3c2adf446025452bcf74>
>
> Indeed, running the test without further change I got the following
> exception similar to your
>
> [ERROR] Errors:
> [ERROR]   TestConfigurationPropertySource.testListValueInjection » UnsatisfiedDependency Error creating bean with name 'org.apache.commons.configuration2.spring.TestConfigurationPropertySource': Unsatisfied dependency expressed through field 'booleanNullValue'; nested exception is org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'boolean'; nested exception is java.lang.IllegalArgumentException: Invalid boolean value []
> [ERROR]   TestConfigurationPropertySource.testNullValueInjection » UnsatisfiedDependency Error creating bean with name 'org.apache.commons.configuration2.spring.TestConfigurationPropertySource': Unsatisfied dependency expressed through field 'booleanNullValue'; nested exception is org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'boolean'; nested exception is java.lang.IllegalArgumentException: Invalid boolean value []
> [ERROR]   TestConfigurationPropertySource.testSystemPropertyValueInjection » UnsatisfiedDependency Error creating bean with name 'org.apache.commons.configuration2.spring.TestConfigurationPropertySource': Unsatisfied dependency expressed through field 'booleanNullValue'; nested exception is org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'boolean'; nested exception is java.lang.IllegalArgumentException: Invalid boolean value []
> [ERROR]   TestConfigurationPropertySource.testValueInjection » UnsatisfiedDependency Error creating bean with name 'org.apache.commons.configuration2.spring.TestConfigurationPropertySource': Unsatisfied dependency expressed through field 'booleanNullValue'; nested exception is org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'boolean'; nested exception is java.lang.IllegalArgumentException: Invalid boolean value []
> [INFO]
> [ERROR] Tests run: 2942, Failures: 0, Errors: 4, Skipped: 2
>
> that are solved with the commit ab388c3
> <https://github.com/apache/commons-configuration/commit/ab388c3a0c97cd249dfc2945b149da8b34e01da6>
>
> I would appreciate a new check on your side. Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-configuration/pull/425#issuecomment-2133052890>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBG2EVEMPE6U3HNK6SFQDXLZEL3XNAVCNFSM6AAAAABID3UVXCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMZTGA2TEOBZGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

garydgregory commented:
All parties seem to agree. I will merge. TY all for your patience.
